### PR TITLE
[FluidDynamicsApp] Using list of double and component variables in Dist. Modification Process

### LINKED ISSUE #4834: [FluidDynamics] Historical variables in DistanceModificationProcess
Hi @KratosMultiphysics/fluid-dynamics!

I'm writing this issue after https://github.com/KratosMultiphysics/Kratos/pull/4811#discussion_r281951267.

In my application (PotentialFlow), I'm using the distance modification process to solve embedded geometries. However, if the "deactivate_fully_negative_elements" option is activated, it is required to define the VELOCITY and PRESSURE as historical variables in the solver, even if you dont need them. Would it be possible to make this optional?

https://github.com/KratosMultiphysics/Kratos/blob/c030f918a296b0b87c6a0c33bf823060791c2f4e/applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp#L352-L363

Also, is there a specific reason that the Historial NodalH process is used intead of the NonHistorical one? 

https://github.com/KratosMultiphysics/Kratos/blob/c030f918a296b0b87c6a0c33bf823060791c2f4e/applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp#L140-L146

Regards!

### PR DESCRIPTION
**Pull request by marcnunezc**
Hi,

This is an attempt to close #4834. Now there's a list of double variables and a list for component variables to be fixed/freed, so the user can select which ones to use or select none. By default PRESSURE and VELOCITY_i components are selected.

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I would suggest a single list to simplify usability, you can always check if the input is a double variable or a component using `KratosComponents<Variable<double>>::Has(name)` and `KratosComponents<ComponentType>::Has(name)`.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Do not access KratosComponents inside a node loop if you can avoid it, it is a relatively expensive call. I propose that you store list of variables instead of list of strings in a separate comment but, even if you don't do that, call KratosComponents once, outside the loop.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Move access to KratosComponents outside the loop (see comment for Variable<double>).

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Not inside a loop if you can avoid it.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Not inside a loop if you can avoid it.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.h
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.h]: I would suggest storing those as `std::vector<const Variable<double>&>` and `std::vector<const ComponentType&>` (or `*` if the reference does not work) and initializing the vector at construction from the string input.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I would suggest initializing vectors of variables (or references/pointers of variables) from the string-based input. This allows you to validate the input at construction (if KratosComponents does not have the variable, you will know it here) and minimizes (expensive) calls to KratosComponents<...>::Get().

**Review by jcotela — CHANGES_REQUESTED** [review]
Good idea, but I would suggest working on the variables internally (instead of using strings with the variable names). This would eliminate a potential performance bottleneck with KratosComponents.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Ok I see!

File: applications/CompressiblePotentialFlowApplication/python_scripts/level_set_remeshing_process.py
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/python_scripts/level_set_remeshing_process.py]: This change should be in another PR probably, but since the change in the DistModProcess "breaks" the tests in the PotentialFlowApp (since VELOCITY/PRESSURE are not added as DOFs), I'd rather do this change  here

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Is the name ok?

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: @jcotela, now there is no "KratosComponents" call in the nodal loop. 

Would you also move the definition of r_double_var and r_component_var outside the nodal loop?

I kept them inside since doing it outside would require to do as many nodal loops as variable in the list, but Im not quite sure which option is faster.

**Review by marcnunezc — COMMENTED** [review]
@jcotela  I implemented your suggestions! I have some questions that I have formulated in the corresponding lines of the code.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I think it is not very descriptive. As an external user, I have no clue of why you need these variables in a process "for distance modification". However, I do not have a better suggestion. @rubenzorrilla any ideas?

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I would do it like this, I think it is better to loop less times over the nodes.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Some more suggestions:
"list_of_dofs_to_fix"
"list_of_variables_to_fix"
"variables_to_fix_list"
"dofs_to_fix_list"
"variables_to_fix"
"dofs_to_fix"

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Very minor but, if you do this before the check, you can avoid getting the variable twice. Also, since you are checking (:+1:, good idea), you could also add `KRATOS_CHECK_VARIABLE_IN_NODAL_DATA`

Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Also, I know I suggested the pointer, but using the value would be much cleaner. The only thing keeping me from recommending that is that we have had some recent issues with copying variables, but feel free to give it a try if you want to.

Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Do you mean using references  `std::vector<const Variable<double>&> `?

Or the variable as it is  `std::vector<const Variable<double>> `?

Using references seems to not work.

Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I mean the second option. References would be the simplest, but they do not play nice with std::vector.

Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Anyways, I would not hold the PR for this. Give it a try if you'd like, but otherwise we go ahead with whatever works.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Since this only applies for the `"deactivate_full_negative_elements": true` case, I'd rather call it `full_negative_elements_fixed_variables_list` or something similar that clearly states that these two arguments are tied.

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: As a comment aside, I don't really like to specify each component one by one since all of them will be always fixed. However, no alternative comes to my mind (apart of creating a vector variables list, which I found it worse).

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Why do you need the cast in here?

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Same, why the cast?

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]:
```suggestion
                    const auto &r_component_var = *mComponentVariablesList[i_comp];
```

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]:
```suggestion
                    const auto &r_double_var = *mDoubleVariablesList[i_var];
```

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]:
```suggestion
            auto it_node = rNodes.begin() + i_node;
```

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]:
```suggestion
                    const auto &r_double_var = *mDoubleVariablesList[i_var];
```

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]:
```suggestion
                    const auto &r_component_var = *mComponentVariablesList[i_comp];
```

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Not long ago I did experience some issues with the `std::vector<const Variable<double>>` so I'd go for the pointer option. However, I'm not so much in favour of raw pointers...

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.h
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.h:220]: Would you mind to add the Doxygen documentation to this new method?

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
Minor comments to be clarified before merge.

**Comment by rubenzorrilla** [comment]
> Minor comments to be clarified before merge.

Furthermore, Travis is crashing

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I dont need it actually, I thought I did since it was inside the OMP loop.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Same as above

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: ok!

**Comment by marcnunezc** [comment]
> > Minor comments to be clarified before merge.
> 
> Furthermore, Travis is crashing

Yes! I have to update the tests! I didnt realize the distance modification process tests were failing since the [run_cpp_unit_test.py](https://github.com/KratosMultiphysics/Kratos/blob/279f9c462826eaacfcb36f855a9c6915eccb7dd5/applications/FluidDynamicsApplication/tests/run_cpp_unit_tests.py#L6) file only runs the sensitivtiy tests. Is this intended?

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: The parallel loop is the outer one. You can use an `unsigned int` iterator in here. Actually, you can even use an auto loop.

**Comment by rubenzorrilla** [comment]
> > > Minor comments to be clarified before merge.
> > 
> > 
> > Furthermore, Travis is crashing
> 
> Yes! I have to update the tests! I didnt realize the distance modification process tests were failing since the [run_cpp_unit_test.py](https://github.com/KratosMultiphysics/Kratos/blob/279f9c462826eaacfcb36f855a9c6915eccb7dd5/applications/FluidDynamicsApplication/tests/run_cpp_unit_tests.py#L6) file only runs the sensitivtiy tests. Is this intended?

In principle it is not intended. I think we merged this script after the tests were created and we missed to add them.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I went for `full_negative_elements_fixed_variables_list` in the end.

Regarding your last comment, I didnt like it either so I added the possibility to define the whole vector, and then we check each component one by one.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Here I check if its a vector variable and then I get each component with Kratos components. Do you like this approach?

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Look much better! Thanks @marcnunezc !

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: This `distance_factor` is here because of legacy reasons (we used to use it in our very first implementation). I search for it and there is only two concurrencies (one in this process and one in its Python apply one).

Maybe it's a good idea to take the chance to remove it in this PR.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: To check `rVariableStringArray[i_variable]` must be enough. You don't need to go through each component.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I see!

I also wanted to switch the NodalH calculation from historical to non historical (was thinking to do so in another PR), would you do it here as well? 

@jcotela, is what you did in #4850 enough to define the nodalH calculation to NonHistorical safely?

Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Yes, but I did not change existing code, so you still need to call the new function wherever appropriate. Also, if you find other missing MPI operations, please let me know and I'll look into adding them.

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: > I see!
> 
> I also wanted to switch the NodalH calculation from historical to non historical (was thinking to do so in another PR), would you do it here as well?
> 
> @jcotela, is what you did in #4850 enough to define the nodalH calculation to NonHistorical safely?

This I'd do it in another PR since it is a potential source of errors.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: The variable in nodal data can be checked just once for the `array_1d<double,3>`, but DOF are added by component (also note that the z component will likely NOT be a dof in 3D)

Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: > (also note that the z component will likely NOT be a dof in 3D)

Did you mean in 2D? Would you add a condition to check if we are in 2D or 3D?

Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Yes, I meant in 2D. I'm afraid we need to check for domain size (probably just get it from the modelpart?) and only use the Z component for 3D.

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: > The variable in nodal data can be checked just once for the `array_1d<double,3>`, but DOF are added by component (also note that the z component will likely NOT be a dof in 3D)

I thought it was a nodal variable :man_facepalming:  forget about it.

**Comment by marcnunezc** [comment]
Hi @rubenzorrilla and @jcotela 

I implemented your last comments: removing distance_factor parameter, using Z component only in 3D and checking the nodal data for the vector variable (for instance checking VELOCITY instead of each of the components separately). DOFs are checked component by component.

**Comment by rubenzorrilla** [comment]
@marcnunezc Codacy is complaining about an unused variable. Aside of this, I've no more comments.

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: @rubenzorrilla , the variable codacy is complaining about is used in this macro to check that the variable exists in the nodal data (it is only used here). Seems like codacy is not "seeing" the Kratos macros?

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: OK then.

**Review by rubenzorrilla — DISMISSED** [review]

**Comment by marcnunezc** [comment]
I also need your approval @jcotela

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: What is `if` checking? If `size() == 0` the `for` loop in the next line should do 0 iterations.

Comment (marcnunezc) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: Yes you are right, the for loop would do 0 iterations, i can remove it.

There is no difference in performance right?

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (marcnunezc) [inline applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp:455]: In the check below I was doing the same as here, but here we are avoiding the nodal loop if both variable list have size()==0.0

**Comment by marcnunezc** [comment]
I removed the unnecessary check @jcotela

File: applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp
Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_processes/distance_modification_process.cpp]: I'd not expect to see a big difference, but in general it is a good practice to reduce branching (so the less if/else, the better).

**Review by jcotela — APPROVED** [review]
:+1: I think all concerns have been addressed. Thanks for your patience.

**Pull request merged**
